Status API

mre opened the pull request:
This is a new endpoint, which returns the timestamps of the latest endpoint calls per account, e.g.

```json
{
    "account_id": 1,
    "latestUpdates": [
        "aggregate": "2021-01-01 00:00:00"
        "detailedStreams": "2021-01-01 00:00:00"
    ]
}
```

It is helpful for building an API status page in the future.

woolfg commented on src/db/StatusRepository.ts:
```diff
@@ -0,0 +1,34 @@
+    // updates for the given accountId and endpoint.
+    // (Updates are stored in JSON format)
+    async getStatus(accountId: number, endpoint: string): Promise<Status> {
+        const query = `SELECT * FROM updates WHERE account_id = ? AND endpoint = ? ORDER BY updated_at DESC LIMIT 1`
```
`ORDER BY timestamp` and you have to add a WHERE clause using the timestamp to be able to make use of the index and do not trigger a full table scan. I would recommend to specify timestamp > DATE_SUB(NOW(), INTERVAL 7 DAY) or something like this

mre replied:
done

mre commented:
> what about just showing the podcast_ids an the last action of it?

We would leak all podcast ids and the API could be a public endpoint, so I'm not sure if we can/should do this. 🤔

mre commented:
Most likely this endpoint will be used by a status page for a single podcast, so providing the token shouldn't be a big deal.

mre commented:
We agreed that we will use MySQL's event scheduler to remove events older than 7 days.

mre commented:
Unfortunately the `event` keyword is not supported by PlanetScale.
See https://planetscale.com/docs/reference/mysql-compatibility and https://github.com/planetscale/discussion/discussions/265.

woolfg commented:
can be testen locally using the `stack` repo?

woolfg commented:
with the new changed in `stack` it should be possible to test it there. according to https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_event_scheduler it is ON by default anyways.

woolfg commented on README.md:
```diff
@@ -39,6 +39,31 @@ It stores one unique vote for each episode per IP and user-agent pair.
+
+To query it, use the following endpoint:
+
+`/status/:accountID`
```
is already defined by the auth token, right?

mre replied:
True. Fixed.

woolfg commented on db_schema/schema.sql:
```diff
@@ -212,4 +212,21 @@ CREATE TABLE IF NOT EXISTS feedbackComment (
+-- automatically delete all updates older than 7 days
+CREATE EVENT IF NOT EXISTS updates_cleanup
+ON SCHEDULE EVERY 1 DAY
+DO DELETE FROM updates WHERE timestamp < (CURRENT_TIMESTAMP - INTERVAL 7 DAY);
```
`DATE_SUB(NOW(), INTERVAL 7 DAY)`

mre replied:
nice

woolfg commented on db_schema/schema.sql:
```diff
@@ -212,4 +212,21 @@ CREATE TABLE IF NOT EXISTS feedbackComment (
+CREATE TABLE updates (
+  account_id INTEGER NOT NULL,
+  endpoint VARCHAR(64) NOT NULL,
+  timestamp TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
wouldn't recommend to use keywords as field names (timestamp TIMESTAMP), maybe just `created`?

woolfg reviewed (approved):
looks good. added some minor remarks. in general it would be nicer to be able to check the status without auth request and just see a table of account_ids, but would merge first and improce later.

mre commented:
* Removed authentication
* Changed response to reply with updates of all accounts.
* Renamed `timestamp` field to `created`
* Updated the field name in PlanetScale (dev only).
* Tested locally

Example response:

```json
{
  "1": {
    "latestUpdates": {
      "metadata": "2022-12-23T15:03:47.000Z",
      "detailedStreams": "2022-12-23T15:05:14.000Z",
      "listeners": "2022-12-23T15:05:17.000Z",
      "aggregate": "2022-12-23T15:05:12.000Z",
      "followers": "2022-12-23T15:04:00.000Z",
      "episodes": "2022-12-23T15:04:03.000Z",
      "performance": "2022-12-23T15:05:19.000Z"
    }
  },
  "2": {
    "latestUpdates": {
      "listeners": "2023-01-30T11:30:38.000Z"
    }
  }
}
```
